src/ai-did-you-mean-this/azext_ai_did_you_mean_this/_telemetry.py

```diff
@@ -0,0 +1,210 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+from functools import wraps
+from typing import Dict, Union, Any
+from enum import Enum, auto
+import azure.cli.core.telemetry as telemetry
+
+from azext_ai_did_you_mean_this._const import (
+    UNEXPECTED_ERROR_STR,
+    EXTENSION_NAME,
+    TELEMETRY_PROPERTY_PREFIX
+)
+
+
+def _user_agrees_to_telemetry(func):
+    @wraps(func)
+    def _wrapper(*args, **kwargs):
+        if not telemetry.is_telemetry_enabled():
+            return None
+
+        return func(*args, **kwargs)
+
+    return _wrapper
+
+
+class FaultType(Enum):
+    RequestError = 'thoth-aladdin-request-error'
+    ParseError = 'thoth-aladdin-response-parse-error'
+    UnexpectedError = 'thoth-unexpected-error'
+
+    def __eq__(self, value: Union['FaultType', str]):
+        if hasattr(value, 'value'):
+            value = value.value
+        # pylint: disable=comparison-with-callable
+        return self.value == value
+
+    def __hash__(self):
+        return hash(self.value)
+
+
+class NoRecommendationReason(Enum):
+    ServiceRequestFailure = 'service-request-failure'
+    CommandFromExtension = 'command-from-extension'
+    EmptyCommand = 'empty-command'
+
+    def __eq__(self, value: Union['NoRecommendationReason', str]):
+        if hasattr(value, 'value'):
+            value = value.value
+        # pylint: disable=comparison-with-callable
+        return self.value == value
+
+    def __hash__(self):
+        return hash(self.value)
+
+
+class TelemetryProperty(Enum):
```

Comment:
curious, where to read your ai key?

Reply:
Addressed comment in offline conversation.